src/boot/efi/xbootldr.c

```diff
@@ -16,31 +21,163 @@ static EFI_DEVICE_PATH *path_parent(EFI_DEVICE_PATH *path, EFI_DEVICE_PATH *node
         CopyMem((UINT8*) parent + len, EndDevicePath, sizeof(EFI_DEVICE_PATH));
 
         return parent;
 }
 
-VOID xbootldr_open(EFI_HANDLE *device, EFI_HANDLE *ret_device, EFI_FILE **ret_root_dir) {
-        EFI_DEVICE_PATH *partition_path, *disk_path, *copy;
-        UINT32 found_partition_number = UINT32_MAX;
-        UINT64 found_partition_start = UINT64_MAX;
-        UINT64 found_partition_size = UINT64_MAX;
-        UINT8 found_partition_signature[16] = {};
-        EFI_HANDLE new_device;
-        EFI_FILE *root_dir;
-        EFI_STATUS r;
+static BOOLEAN verify_gpt(union GptHeaderBuffer *gpt_header_buffer, EFI_LBA lba_expected) {
+        EFI_PARTITION_TABLE_HEADER *h;
+        UINT32 crc32, crc32_saved;
+        EFI_STATUS err;
+
+        assert(gpt_header_buffer);
+
+        h = &gpt_header_buffer->gpt_header;
+
+        /* Some superficial validation of the GPT header */
+        if(CompareMem(&h->Header.Signature, "EFI PART", sizeof(h->Header.Signature) != 0))
+                return FALSE;
+
+        if (h->Header.HeaderSize < 92 || h->Header.HeaderSize > 512)
+                return FALSE;
+
+        if (h->Header.Revision != 0x00010000U)
+                return FALSE;
+
+        /* Calculate CRC check */
+        crc32_saved = h->Header.CRC32;
+        h->Header.CRC32 = 0;
+        err = BS->CalculateCrc32(gpt_header_buffer, h->Header.HeaderSize, &crc32);
+        h->Header.CRC32 = crc32_saved;
+        if (EFI_ERROR(err) || crc32 != crc32_saved)
+                return FALSE;
+
+        if (h->MyLBA != lba_expected)
+                return FALSE;
+
+        if (h->SizeOfPartitionEntry < sizeof(EFI_PARTITION_ENTRY))
+                return FALSE;
+
+        if (h->NumberOfPartitionEntries <= 0 || h->NumberOfPartitionEntries > 1024)
+                return FALSE;
+
+        /* overflow check */
+        if (h->SizeOfPartitionEntry > UINTN_MAX / h->NumberOfPartitionEntries)
+                return FALSE;
+
+        return TRUE;
+}
+
+static EFI_STATUS try_gpt(
+                EFI_BLOCK_IO *block_io,
+                EFI_LBA lba,
+                UINT32 *ret_part_number,
+                UINT64 *ret_part_start,
+                UINT64 *ret_part_size,
+                EFI_GUID *ret_part_uuid) {
+
+        _cleanup_freepool_ EFI_PARTITION_ENTRY *entries = NULL;
+        union GptHeaderBuffer gpt;
+        EFI_STATUS err;
+        UINT32 crc32;
+        UINTN size;
+
+        assert(block_io);
+        assert(ret_part_number);
+        assert(ret_part_start);
+        assert(ret_part_size);
+        assert(ret_part_uuid);
+
+        /* Read the GPT header */
+        err = uefi_call_wrapper(
+                        block_io->ReadBlocks, 5,
+                        block_io,
+                        block_io->Media->MediaId,
+                        lba,
+                        sizeof(gpt), &gpt);
+        if (EFI_ERROR(err))
+                return err;
+
+        if (!verify_gpt(&gpt, lba))
+                return EFI_DEVICE_ERROR;
```

Comment:
that sounds like an I/O error, i.e. hw error or so. Maybe just rais EFI_NOT_FOUND here? sounds more appropriate to me.

(but of course, doesn't matter